java/cdk/src/main/java/sleeper/cdk/stack/CompactionStack.java

```diff
@@ -439,9 +526,108 @@ private Cluster ecsClusterForSplittingCompactionTasks(IBucket configBucket,
         return cluster;
     }
 
-    private FargateTaskDefinition compactionFargateTaskDefinition(String compactionTypeName) {
+    private void addEC2CapacityProvider(Cluster cluster, String clusterName, IVpc vpc,
+            SystemDefinedInstanceProperty scalingProperty) {
+        AutoScalingGroup ec2scalingGroup = AutoScalingGroup.Builder.create(this, clusterName + "ScalingGroup").vpc(vpc)
+                .allowAllOutbound(true)
+                .associatePublicIpAddress(false)
+                .requireImdsv2(true)
+                .blockDevices(Arrays.asList(BlockDevice.builder()
+                        .deviceName("/dev/xvda") // root volume
+                        .volume(BlockDeviceVolume.ebs(instanceProperties.getInt(COMPACTION_EC2_ROOT_SIZE),
+                                EbsDeviceOptions.builder()
+                                        .deleteOnTermination(true)
+                                        .encrypted(true)
+                                        .volumeType(EbsDeviceVolumeType.GP2)
+                                        .build()))
+                        .build()))
+                .minCapacity(instanceProperties.getInt(COMPACTION_EC2_POOL_MINIMUM))
+                .desiredCapacity(instanceProperties.getInt(COMPACTION_EC2_POOL_DESIRED))
+                .maxCapacity(instanceProperties.getInt(COMPACTION_EC2_POOL_MAXIMUM)).requireImdsv2(true)
+                .instanceType(lookupEC2InstanceType(instanceProperties.get(COMPACTION_EC2_TYPE)))
+                .machineImage(EcsOptimizedImage.amazonLinux2(AmiHardwareType.GPU,
+                        EcsOptimizedImageOptions.builder()
+                                .cachedInContext(false)
+                                .build()))
+                .build();
 
-        String architecture = instanceProperties.get(COMPACTION_TASK_CPU_ARCHITECTURE).toUpperCase(Locale.ROOT);
+        AsgCapacityProvider ec2Provider = AsgCapacityProvider.Builder
+                .create(this, clusterName + "CapacityProvider")
+                .enableManagedScaling(false)
+                .enableManagedTerminationProtection(false)
+                .autoScalingGroup(ec2scalingGroup)
+                .spotInstanceDraining(true)
+                .canContainersAccessInstanceRole(false)
+                .machineImageType(MachineImageType.AMAZON_LINUX_2)
+                .build();
+
+        cluster.addAsgCapacityProvider(ec2Provider,
+                AddAutoScalingGroupCapacityOptions.builder()
+                        .canContainersAccessInstanceRole(false)
+                        .machineImageType(MachineImageType.AMAZON_LINUX_2)
+                        .spotInstanceDraining(true)
+                        .build());
+
+        instanceProperties.set(scalingProperty, ec2scalingGroup.getAutoScalingGroupName());
+    }
+
+    public static InstanceType lookupEC2InstanceType(String ec2InstanceType) {
```

Comment:
Should we use the constructor InstanceType(String instanceTypeIdentifier)?

Reply:
InstanceType doesn't have a public constructor that I can see. Are you looking at a different class of the same name?

Reply:
No, there's a public InstanceType(String instanceTypeIdentifier). That's on software.amazon.awscdk.services.ec2.InstanceType, in aws-cdk-lib:2.35.0. My IDE says that's the class that's used here.

Reply:
I can see the constructor in the source code, but it isn't in the public API for some reason. Although the constructor itself is public. I'm not sure if we should use the static. InstanceType.of, or call the constructor directly with the string.